File: native/ops/comparison.c
/**
 * Real + complex comparison ops.  Output is a logical tensor stored as
 * 0.0 / 1.0 in a double buffer (numbl stores logicals as FloatXArray with
 * an _isLogical flag on the runtime tensor).
 *
 * Caller-allocated input/output buffers; never copies.
 */

#include "numbl_ops.h"

static inline double rcmp(int op, double a, double b) {
  switch (op) {
    case NUMBL_CMP_EQ: return a == b ? 1.0 : 0.0;
    case NUMBL_CMP_NE: return a != b ? 1.0 : 0.0;
    case NUMBL_CMP_LT: return a <  b ? 1.0 : 0.0;
    case NUMBL_CMP_LE: return a <= b ? 1.0 : 0.0;
    case NUMBL_CMP_GT: return a >  b ? 1.0 : 0.0;
    case NUMBL_CMP_GE: return a >= b ? 1.0 : 0.0;
    default:           return -1.0;   /* sentinel; caller checks bad op */
  }
}

int numbl_real_comparison(int op, size_t n,
                          const double* a, const double* b,
                          double* out) {
  if (!a || !b || !out) return NUMBL_ERR_NULL_PTR;
  if (op < 0 || op > NUMBL_CMP_GE) return NUMBL_ERR_BAD_OP;
  switch (op) {
    case NUMBL_CMP_EQ:
      for (size_t i = 0; i < n; i++) out[i] = a[i] == b[i] ? 1.0 : 0.0;
      return NUMBL_OK;
    case NUMBL_CMP_NE:
      for (size_t i = 0; i < n; i++) out[i] = a[i] != b[i] ? 1.0 : 0.0;
      return NUMBL_OK;
    case NUMBL_CMP_LT:
      for (size_t i = 0; i < n; i++) out[i] = a[i] <  b[i] ? 1.0 : 0.0;
      return NUMBL_OK;
    case NUMBL_CMP_LE:
      for (size_t i = 0; i < n; i++) out[i] = a[i] <= b[i] ? 1.0 : 0.0;
      return NUMBL_OK;
    case NUMBL_CMP_GT:
      for (size_t i = 0; i < n; i++) out[i] = a[i] >  b[i] ? 1.0 : 0.0;
      return NUMBL_OK;
    case NUMBL_CMP_GE:
      for (size_t i = 0; i < n; i++) out[i] = a[i] >= b[i] ? 1.0 : 0.0;
      return NUMBL_OK;
    default:
      return NUMBL_ERR_BAD_OP;
  }
}

int numbl_real_scalar_comparison(int op, size_t n,
                                 double scalar, const double* arr,
                                 int scalar_on_left, double* out) {
  if (!arr || !out) return NUMBL_ERR_NULL_PTR;
  if (op < 0 || op > NUMBL_CMP_GE) return NUMBL_ERR_BAD_OP;
  if (scalar_on_left) {
    for (size_t i = 0; i < n; i++) out[i] = rcmp(op, scalar, arr[i]);
  } else {
    for (size_t i = 0; i < n; i++) out[i] = rcmp(op, arr[i], scalar);
  }
  return NUMBL_OK;
}

int numbl_complex_comparison(int op, size_t n,
                             const double* a_re, const double* a_im,
                             const double* b_re, const double* b_im,
                             double* out) {
  if (!a_re || !b_re || !out) return NUMBL_ERR_NULL_PTR;
  switch (op) {
    case NUMBL_CMP_EQ:
      for (size_t i = 0; i < n; i++) {
        double ar = a_re[i], ai = a_im ? a_im[i] : 0.0;
        double br = b_re[i], bi = b_im ? b_im[i] : 0.0;
        out[i] = (ar == br && ai == bi) ? 1.0 : 0.0;
      }
      return NUMBL_OK;
    case NUMBL_CMP_NE:
      for (size_t i = 0; i < n; i++) {
        double ar = a_re[i], ai = a_im ? a_im[i] : 0.0;
        double br = b_re[i], bi = b_im ? b_im[i] : 0.0;
        out[i] = (ar != br || ai != bi) ? 1.0 : 0.0;
      }
      return NUMBL_OK;
    /* MATLAB semantics: <, <=, >, >= compare real parts only. */
    case NUMBL_CMP_LT:
      for (size_t i = 0; i < n; i++) out[i] = a_re[i] <  b_re[i] ? 1.0 : 0.0;
      return NUMBL_OK;
    case NUMBL_CMP_LE:
      for (size_t i = 0; i < n; i++) out[i] = a_re[i] <= b_re[i] ? 1.0 : 0.0;
      return NUMBL_OK;
    case NUMBL_CMP_GT:
      for (size_t i = 0; i < n; i++) out[i] = a_re[i] >  b_re[i] ? 1.0 : 0.0;
      return NUMBL_OK;
    case NUMBL_CMP_GE:
      for (size_t i = 0; i < n; i++) out[i] = a_re[i] >= b_re[i] ? 1.0 : 0.0;
      return NUMBL_OK;
    default:
      return NUMBL_ERR_BAD_OP;
  }
}

int numbl_complex_scalar_comparison(int op, size_t n,
                                    double s_re, double s_im,
                                    const double* arr_re,
                                    const double* arr_im,
                                    int scalar_on_left, double* out) {
  if (!arr_re || !out) return NUMBL_ERR_NULL_PTR;
  switch (op) {
    case NUMBL_CMP_EQ:
      if (scalar_on_left) {
        for (size_t i = 0; i < n; i++) {
          double ar = arr_re[i], ai = arr_im ? arr_im[i] : 0.0;
          out[i] = (s_re == ar && s_im == ai) ? 1.0 : 0.0;
        }
      } else {
        for (size_t i = 0; i < n; i++) {
          double ar = arr_re[i], ai = arr_im ? arr_im[i] : 0.0;
          out[i] = (ar == s_re && ai == s_im) ? 1.0 : 0.0;
        }
      }
      return NUMBL_OK;
    case NUMBL_CMP_NE:
      if (scalar_on_left) {
        for (size_t i = 0; i < n; i++) {
          double ar = arr_re[i], ai = arr_im ? arr_im[i] : 0.0;
          out[i] = (s_re != ar || s_im != ai) ? 1.0 : 0.0;
        }
      } else {
        for (size_t i = 0; i < n; i++) {
          double ar = arr_re[i], ai = arr_im ? arr_im[i] : 0.0;
          out[i] = (ar != s_re || ai != s_im) ? 1.0 : 0.0;
        }
      }
      return NUMBL_OK;
    /* Real-part-only comparisons. */
    case NUMBL_CMP_LT:
    case NUMBL_CMP_LE:
    case NUMBL_CMP_GT:
    case NUMBL_CMP_GE:
      if (scalar_on_left) {
        for (size_t i = 0; i < n; i++) out[i] = rcmp(op, s_re, arr_re[i]);
      } else {
        for (size_t i = 0; i < n; i++) out[i] = rcmp(op, arr_re[i], s_re);
      }
      return NUMBL_OK;
    default:
      return NUMBL_ERR_BAD_OP;
  }
}

File: src/numbl-core/ops/comparison.ts
/**
 * Pure-TS implementation of comparison ops.
 * Mirrors native/ops/comparison.c.
 *
 * Output is 0.0 / 1.0 stored in a Float64Array (numbl's logical-tensor
 * convention — caller wraps the result with _isLogical = true).
 */

import { OpCmp } from "./opCodes.js";

function rcmp(op: number, a: number, b: number): number {
  switch (op) {
    case OpCmp.EQ:
      return a === b ? 1 : 0;
    case OpCmp.NE:
      return a !== b ? 1 : 0;
    case OpCmp.LT:
      return a < b ? 1 : 0;
    case OpCmp.LE:
      return a <= b ? 1 : 0;
    case OpCmp.GT:
      return a > b ? 1 : 0;
    case OpCmp.GE:
      return a >= b ? 1 : 0;
    default:
      throw new Error(`tsRealComparison: unknown op ${op}`);
  }
}

export function tsRealComparison(
  op: number,
  n: number,
  a: Float64Array,
  b: Float64Array,
  out: Float64Array
): void {
  switch (op) {
    case OpCmp.EQ:
      for (let i = 0; i < n; i++) out[i] = a[i] === b[i] ? 1 : 0;
      return;
    case OpCmp.NE:
      for (let i = 0; i < n; i++) out[i] = a[i] !== b[i] ? 1 : 0;
      return;
    case OpCmp.LT:
      for (let i = 0; i < n; i++) out[i] = a[i] < b[i] ? 1 : 0;
      return;
    case OpCmp.LE:
      for (let i = 0; i < n; i++) out[i] = a[i] <= b[i] ? 1 : 0;
      return;
    case OpCmp.GT:
      for (let i = 0; i < n; i++) out[i] = a[i] > b[i] ? 1 : 0;
      return;
    case OpCmp.GE:
      for (let i = 0; i < n; i++) out[i] = a[i] >= b[i] ? 1 : 0;
      return;
    default:
      throw new Error(`tsRealComparison: unknown op ${op}`);
  }
}

export function tsRealScalarComparison(
  op: number,
  n: number,
  scalar: number,
  arr: Float64Array,
  scalarOnLeft: boolean,
  out: Float64Array
): void {
  if (scalarOnLeft) {
    for (let i = 0; i < n; i++) out[i] = rcmp(op, scalar, arr[i]);
  } else {
    for (let i = 0; i < n; i++) out[i] = rcmp(op, arr[i], scalar);
  }
}

export function tsComplexComparison(
  op: number,
  n: number,
  aRe: Float64Array,
  aIm: Float64Array | null,
  bRe: Float64Array,
  bIm: Float64Array | null,
  out: Float64Array
): void {
  switch (op) {
    case OpCmp.EQ:
      for (let i = 0; i < n; i++) {
        const ar = aRe[i],
          ai = aIm ? aIm[i] : 0;
        const br = bRe[i],
          bi = bIm ? bIm[i] : 0;
        out[i] = ar === br && ai === bi ? 1 : 0;
      }
      return;
    case OpCmp.NE:
      for (let i = 0; i < n; i++) {
        const ar = aRe[i],
          ai = aIm ? aIm[i] : 0;
        const br = bRe[i],
          bi = bIm ? bIm[i] : 0;
        out[i] = ar !== br || ai !== bi ? 1 : 0;
      }
      return;
    // MATLAB semantics: ordering ops use real parts only.
    case OpCmp.LT:
      for (let i = 0; i < n; i++) out[i] = aRe[i] < bRe[i] ? 1 : 0;
      return;
    case OpCmp.LE:
      for (let i = 0; i < n; i++) out[i] = aRe[i] <= bRe[i] ? 1 : 0;
      return;
    case OpCmp.GT:
      for (let i = 0; i < n; i++) out[i] = aRe[i] > bRe[i] ? 1 : 0;
      return;
    case OpCmp.GE:
      for (let i = 0; i < n; i++) out[i] = aRe[i] >= bRe[i] ? 1 : 0;
      return;
    default:
      throw new Error(`tsComplexComparison: unknown op ${op}`);
  }
}

export function tsComplexScalarComparison(
  op: number,
  n: number,
  sRe: number,
  sIm: number,
  arrRe: Float64Array,
  arrIm: Float64Array | null,
  scalarOnLeft: boolean,
  out: Float64Array
): void {
  switch (op) {
    case OpCmp.EQ:
      if (scalarOnLeft) {
        for (let i = 0; i < n; i++) {
          const ar = arrRe[i],
            ai = arrIm ? arrIm[i] : 0;
          out[i] = sRe === ar && sIm === ai ? 1 : 0;
        }
      } else {
        for (let i = 0; i < n; i++) {
          const ar = arrRe[i],
            ai = arrIm ? arrIm[i] : 0;
          out[i] = ar === sRe && ai === sIm ? 1 : 0;
        }
      }
      return;
    case OpCmp.NE:
      if (scalarOnLeft) {
        for (let i = 0; i < n; i++) {
          const ar = arrRe[i],
            ai = arrIm ? arrIm[i] : 0;
          out[i] = sRe !== ar || sIm !== ai ? 1 : 0;
        }
      } else {
        for (let i = 0; i < n; i++) {
          const ar = arrRe[i],
            ai = arrIm ? arrIm[i] : 0;
          out[i] = ar !== sRe || ai !== sIm ? 1 : 0;
        }
      }
      return;
    case OpCmp.LT:
    case OpCmp.LE:
    case OpCmp.GT:
    case OpCmp.GE:
      if (scalarOnLeft) {
        for (let i = 0; i < n; i++) out[i] = rcmp(op, sRe, arrRe[i]);
      } else {
        for (let i = 0; i < n; i++) out[i] = rcmp(op, arrRe[i], sRe);
      }
      return;
    default:
      throw new Error(`tsComplexScalarComparison: unknown op ${op}`);
  }
}
